src/frequenz/sdk/actor/power_distributing/_component_managers/_pv_inverter_manager/_pv_inverter_manager.py

```
@@ -0,0 +1,230 @@
+            # greater than 0.0, we can stop allocating further.
+            if remaining_power > Power.zero() or is_close_to_zero(
+                remaining_power.as_watts()
+            ):
+                break
+            distribution = remaining_power / float(num_components - idx)
+            inv_data = self._component_data_caches[inv_id]
+            if not inv_data.has_value():
+                allocations[inv_id] = Power.zero()
+                # Can't get device bounds, so can't use inverter.
+                continue
+            discharge_bounds = Power.from_watts(
+                inv_data.get().active_power_inclusion_lower_bound
+            )
+            # Because all 3 values are negative or 0, we use max, to get the value
+            # with the least absolute value.
+            allocated_power = max(remaining_power, discharge_bounds, distribution)
+            allocations[inv_id] = allocated_power
+            remaining_power -= allocated_power
+
+        _logger.debug(
+            "Distributing %s to PV inverters %s",
+            request.power,
+            allocations,
+        )
+        await self._set_api_power(request, allocations, remaining_power)
+
+    async def _set_api_power(
+        self, request: Request, allocations: dict[int, Power], remaining_power: Power
+    ) -> None:
+        api_client = connection_manager.get().api_client
+        tasks: dict[int, asyncio.Task[None]] = {}
+        for component_id, power in allocations.items():
+            tasks[component_id] = asyncio.create_task(
+                api_client.set_power(component_id, power.as_watts())
+            )
+        _, pending = await asyncio.wait(
+            tasks.values(),
+            timeout=request.request_timeout.total_seconds(),
+            return_when=asyncio.ALL_COMPLETED,
+        )
+        for task in pending:
+            task.cancel()
+            await asyncio.gather(*pending, return_exceptions=True)
```

I was a bit confused by this thinking there could be no pending tasks because the use of `ALL_COMPLETED`, and then it was a big red light to see exceptions being silently swallowed. But they can be pending tasks because of the timeout, right? And you ignore exceptions because you want to kill and ignore the pending stuff, you don't care about exceptions because they should only be `CancelledError` or even if there is a real exception you don't care about it because the task is discarded?

Maybe it is worth an early comment, even when things are much more clear once one keeps reading. It got me thinking about it for a while.

---

oops, looks like an indentation error, fixed.

also, we are not ignoring exceptions, they are processed in later stages.  add a line.

---

> oops, looks like an indentation error, fixed.

I didn't even noticed the indentation error :laughing: